pareto/utilities/bounding_functions.py

```diff
@@ -0,0 +1,170 @@
+#####################################################################################################
+# PARETO was produced under the DOE Produced Water Application for Beneficial Reuse Environmental
+# Impact and Treatment Optimization (PARETO), and is copyright (c) 2021-2023 by the software owners:
+# The Regents of the University of California, through Lawrence Berkeley National Laboratory, et al.
+# All rights reserved.
+#
+# NOTICE. This Software was developed under funding from the U.S. Department of Energy and the U.S.
+# Government consequently retains certain rights. As such, the U.S. Government has been granted for
+# itself and others acting on its behalf a paid-up, nonexclusive, irrevocable, worldwide license in
+# the Software to reproduce, distribute copies to the public, prepare derivative works, and perform
+# publicly and display publicly, and to permit others to do so.
+#####################################################################################################
+# Title: Supporting functions for both Operational and Strategic models
+
+###############################################################################
+### Imports
+from pareto.strategic_water_management.strategic_produced_water_optimization import (
+    WaterQuality,
+)
+from pyomo.environ import (
+    Param,
+    value,
+    Any,
+)
+
+
+###############################################################################
+### Variable bounding manager functions
+def VariableBounds(model):
+    """
+    This function adds bounds to variables in an PARETO existing model object.
+    At present, the following variables are included:
+        > v_F_Piped
+        > v_Q
+        > v_F_Trucked (NOTE: this is to be included in a future revision, but requires a change within the PARETO Strategic model code)
+    """
+    model = Bound_v_F_Piped(model)
+    if model.config.water_quality is WaterQuality.post_process:
+        model = Bound_v_Q_PP(model)
+    elif model.config.water_quality is WaterQuality.discrete:
+        model = Bound_v_F_DiscretePiped(model)
+        # model = Bound_v_F_Trucked(model) # TODO: resolve unit problem with p_max_number_of_trucks - it must be week^-1 to make the units work out
+    elif model.config.water_quality is WaterQuality.false:
+        pass
+    else:
+        pass
+    return model
+
+
+###############################################################################
+### Individual variable bound generation functions
+def Bound_v_F_Piped(model):
+    # initialization function for parameter
+    def p_F_Piped_UB_init(model, l, l_tilde):
+        # set default value to pipeline capacity plus maximum expansion capacity
+        return model.p_sigma_Pipeline[l, l_tilde] + max(
+            [value(model.p_delta_Pipeline[d]) for d in model.s_D]
+        )
```

Comment:
**If the pipeline** connecting any two arcs **is bi-directional** and the model decides to install pipelines in both directions (this may be likely for trunk lines in new development areas), the choice of bound could be limiting the flow (at least conceptually)

Reply:
To my knowledge there is not a "pipe" index that I could use to resolve this problem. That's a PARETO-level limitation as I see it. It won't be solved by these utility functions. I'm marking it as resolved without further consideration.

Reply:
No, that is not correct, you need to adjust the value of the bound otherwise the value will be incorrect. Please review the two constraints in the model `PipelineCapacityExpansion` and `PipelineCapacity`. Although the index for pipeline is not modeled, but we it recognizes multiple pipelines through origin and destination and so, the capacity gets adjusted based on how many pipelines are there between 2 given nodes. An easiest way to do this, may be, is to add a factor of 2 to max() term in your method, i.e., 2 * max(..). A better way would be to identify the bi-directional pipelines and only apply the modification to them, your choice.

Reply:
Good catch, Naresh!